### PR TITLE
Fix Out-of-Bounds Access and Improve Testability in ncio_mod for geoinst_list Array

### PR DESCRIPTION
**Pull request by amstokely**
TYPE: Bug fix

KEYWORDS: bounds checking, stability, modularity, testability

SOURCE: internal

DESCRIPTION OF CHANGES:

Problem: In the ncio_mod module, the geoinst_list array was being accessed out of bounds. This issue was discovered while running the gdas.1bamua.t00z.20180415.bufr calculation in Release mode using GCC 11.4.1. The out-of-bounds access resulted in frequent buffer overflows, leading to potential instability and occasional crashes.

Solution: To resolve this issue and improve code maintainability, the logic for setting observation error flags based on the instrument type has been refactored into a new subroutine, determine_obserr_type. This new subroutine performs the following:

Bounds Checking: Ensures that the ityp index is within bounds before accessing geoinst_list. If ityp is within bounds and geoinst_list(ityp) is 'ahi_himawari8', the set_ahi flag is set to .true.. Otherwise, set_brit is set to .true.
Modularity: Isolating this flag-setting logic into determine_obserr_type allows it to be tested independently, ensuring safe access and consistent behavior.
Benefits:

Improved Stability: The bounds check prevents buffer overflow, enhancing the overall stability and reliability of ncio_mod.
Enhanced Testability: With the logic in a separate subroutine, we can now independently test determine_obserr_type for various geo_inst inputs and ityp values, including edge cases like out-of-bounds indices. This modular design allows for targeted testing without running the full write_obs routine, increasing both the efficiency and coverage of our tests.

### CODE REVIEW

**Comment by amstokely** [comment]
My linter reformated ncio_mod, so make sure to turn on "hide_whitespace" so that the diff isn't the entire file.

**Comment by jim-p-w** [comment]
Is adding the empty `.gitmodules` file intentional?

**Comment by jim-p-w** [comment]
Are the whitespace changes going to put your name as the author of every line in obs2ioda-v2/src/ncio_mod.f90? If so, I'd undo the whitespace changes in that file.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Check CMake version
-cmake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.20)
```
Comment (jim-p-w) [inline CMakeLists.txt:2]: What is in cmake 3.20 that isn't in version 3.10, that is required to build this software?

Comment (amstokely) [reply CMakeLists.txt:2]: In CMake 3.10, if you use a package that sets a CMake variable equal to <PACKAGE_NAME>_ROOT, the variable will be ignored unless you set CMake policy CMP0074. This requirement was deprecated in 3.2.0.  It is pretty common for modern CMake packages to set variables equal to <PACKAGE_NAME>_ROOT, so I figured I should update the minimum cmake version to 3.20,

File: obs2ioda-v2/src/ncio_mod.f90 [hunk 2]
```diff
@@ -1,344 +1,379 @@
...
+    !! @param geo_inst (character(len=*), dimension(:), intent(in)):
+    !!   1D array of instrument names, each entry representing a satellite or instrument type.
+    !! @param ityp (integer(i_kind), intent(in)):
+    !!   Index into the `geo_inst` array specifying which instrument type to check.
+    !! @param set_brit (logical, intent(out)):
+    !!   Output flag set to `.true.` if the instrument at `geo_inst(ityp)` is not 'ahi_himawari8'
+    !!   or if `ityp` is out of bounds.
+    !! @param set_ahi (logical, intent(out)):
+    !!   Output flag set to `.true.` if the instrument at `geo_inst(ityp)` is 'ahi_himawari8'.
+    subroutine determine_obserr_type(geo_inst, ityp, set_brit, set_ahi)
```
[the remark this answers is not in the document]
Comment (amstokely) [reply obs2ioda-v2/src/ncio_mod.f90]: Good point. I'll probably rename it to is_ahi_himawari so that the name will make sense when support for ahi_himawari9 is added.

**Comment by amstokely** [comment]
> Is adding the empty `.gitmodules` file intentional?

No it was not...good catch!

**Review by jim-p-w — CHANGES_REQUESTED** [review]

File: obs2ioda-v2/src/ncio_mod.f90 [hunk 3]
```diff
@@ -1,344 +1,375 @@
...
+            else if (write_opt == write_nc_radiance_geo) then
+                ncfname = trim(outdir) // trim(geoinst_list(ityp)) // '_obs_' // trim(filedate) // '.h5'
+            end if
+            if (write_opt == write_nc_radiance .or. write_opt == write_nc_radiance_geo) then
+                iv = ufo_vars_getindex(name_sen_info, 'sensor_channel')
+                allocate (ichan(xdata(ityp, itim)%nvars))
+                ichan(:) = xdata(ityp, itim)%xseninfo_int(:, iv)
+                allocate (obserr(xdata(ityp, itim)%nvars))
+                if (is_himawari(geoinst_list, "ahi_himawari8", ityp) .eqv. .true.) then
+                    call set_brit_obserr(inst_list(ityp), xdata(ityp, itim)%nvars, obserr)
```
[the remark this answers is not in the document]
Comment (amstokely) [reply obs2ioda-v2/src/ncio_mod.f90]: Again good catch! I moved this logic into it's own subroutine so I can test it.

**Comment by jim-p-w** [comment]
I think all of the unit test bits should be in a separate PR from the change which fixes the out of bounds bug.
The unit test bits stand on their own, and aren't really germane to the out of bounds bug, especially the infrastructure for adding unit tests to the repo.

**Comment by amstokely** [comment]
> Are the whitespace changes going to put your name as the author of every line in obs2ioda-v2/src/ncio_mod.f90? If so, I'd undo the whitespace changes in that file.
> 
> @amstokely Thanks for attending to this, it was prob a pain.

@jim-p-w Took the better part of the morning haha. It is easy to remove the white space, which fixes the diff size problem, but reverting the authorship from a white space change is extremely difficult.

**Comment by mgduda** [comment]
While adding a check that `ityp <= ubound(geo_inst, DIM = 1)` before accessing `geoinst_list(ityp)` will avoid any access to indices above the allocated range of `geoinst_list`, I think it may be worth taking a step back to understand the reason why `ityp` is out of bounds in the first place. In other words, before addressing the symptom, understand the cause.

In `define_mod.f90`, we have:
```fortran
integer(i_kind), parameter :: ninst_geo         = 1
```
```fortran
character(len=nstring), dimension(ninst_geo) :: geoinst_list = &
   (/                     &
      'ahi_himawari8   '  &
   /)
```
and we also have
```fortran
integer(i_kind), parameter :: ninst             = 17
```
```fortran
character(len=nstring), dimension(ninst) :: inst_list = &
   (/                     &
      'amsua_n15       ', &
      'amsua_n18       ', &
      'amsua_n19       ', &
      'amsua_metop-a   ', &
      'amsua_metop-b   ', &
      'amsua_metop-c   ', &
!      'airs_aqua       ', &
      'amsua_aqua      ', &
      'mhs_n18         ', &
      'mhs_n19         ', &
      'mhs_metop-a     ', &
      'mhs_metop-b     ', &
      'mhs_metop-c     ', &
      'iasi_metop-a    ', &
      'iasi_metop-b    ', &
      'iasi_metop-c    ', &
      'cris_npp        ', &
      'cris_n20        '  &
   /)
```
Meanwhile, the logic in `ncio_mod.f90` has:
```fortran
   else if ( write_opt == write_nc_radiance ) then
      ntype = ninst
   else if ( write_opt == write_nc_radiance_geo ) then
      ntype = ninst_geo
```
and later,
```fortran
   obtype_loop: do ityp = 1, ntype
...
      if ( write_opt == write_nc_radiance .or. write_opt == write_nc_radiance_geo ) then
...
         if  ( geoinst_list(ityp) == 'ahi_himawari8' ) then
             call set_ahi_obserr(geoinst_list(ityp), xdata(ityp,itim)%nvars, obserr)
         else
             call set_brit_obserr(inst_list(ityp), xdata(ityp,itim)%nvars, obserr)
         end if
      end if
...
   end do obtype_loop
```
So it seems that we may be accessing `geoinst_list` in cases where `write_opt == write_nc_radiance`, in which case `ntype` is set to be too large. Is it really the intent to access `geoinst_list` when `write_opt == write_nc_radiance_geo` and also when `write_opt == write_nc_radiance`?

**Comment by amstokely** [comment]
> While adding a check that `ityp <= ubound(geo_inst, DIM = 1)` before accessing `geoinst_list(ityp)` will avoid any access to indices above the allocated range of `geoinst_list`, I think it may be worth taking a step back to understand the reason why `ityp` is out of bounds in the first place. In other words, before addressing the symptom, understand the cause.
> 
> In `define_mod.f90`, we have:
> 
> ```fortran
> integer(i_kind), parameter :: ninst_geo         = 1
> ```
> 
> ```fortran
> character(len=nstring), dimension(ninst_geo) :: geoinst_list = &
>    (/                     &
>       'ahi_himawari8   '  &
>    /)
> ```
> 
> and we also have
> 
> ```fortran
> integer(i_kind), parameter :: ninst             = 17
> ```
> 
> ```fortran
> character(len=nstring), dimension(ninst) :: inst_list = &
>    (/                     &
>       'amsua_n15       ', &
>       'amsua_n18       ', &
>       'amsua_n19       ', &
>       'amsua_metop-a   ', &
>       'amsua_metop-b   ', &
>       'amsua_metop-c   ', &
> !      'airs_aqua       ', &
>       'amsua_aqua      ', &
>       'mhs_n18         ', &
>       'mhs_n19         ', &
>       'mhs_metop-a     ', &
>       'mhs_metop-b     ', &
>       'mhs_metop-c     ', &
>       'iasi_metop-a    ', &
>       'iasi_metop-b    ', &
>       'iasi_metop-c    ', &
>       'cris_npp        ', &
>       'cris_n20        '  &
>    /)
> ```
> 
> Meanwhile, the logic in `ncio_mod.f90` has:
> 
> ```fortran
>    else if ( write_opt == write_nc_radiance ) then
>       ntype = ninst
>    else if ( write_opt == write_nc_radiance_geo ) then
>       ntype = ninst_geo
> ```
> 
> and later,
> 
> ```fortran
>    obtype_loop: do ityp = 1, ntype
> ...
>       if ( write_opt == write_nc_radiance .or. write_opt == write_nc_radiance_geo ) then
> ...
>          if  ( geoinst_list(ityp) == 'ahi_himawari8' ) then
>              call set_ahi_obserr(geoinst_list(ityp), xdata(ityp,itim)%nvars, obserr)
>          else
>              call set_brit_obserr(inst_list(ityp), xdata(ityp,itim)%nvars, obserr)
>          end if
>       end if
> ...
>    end do obtype_loop
> ```
> 
> So it seems that we may be accessing `geoinst_list` in cases where `write_opt == write_nc_radiance`, in which case `ntype` is set to be too large. Is it really the intent to access `geoinst_list` when `write_opt == write_nc_radiance_geo` and also when `write_opt == write_nc_radiance`?

@ibanos90 is probably more qualified to answer this than I am.

**Comment by ibanos90** [comment]
> While adding a check that `ityp <= ubound(geo_inst, DIM = 1)` before accessing `geoinst_list(ityp)` will avoid any access to indices above the allocated range of `geoinst_list`, I think it may be worth taking a step back to understand the reason why `ityp` is out of bounds in the first place. In other words, before addressing the symptom, understand the cause.
>
> ...
>
> So it seems that we may be accessing `geoinst_list` in cases where `write_opt == write_nc_radiance`, in which case `ntype` is set to be too large. Is it really the intent to access `geoinst_list` when `write_opt == write_nc_radiance_geo` and also when `write_opt == write_nc_radiance`?

@mgduda, that's a good point and maybe the reason for this issue. We don't need to access `geoinst_list` when `write_opt == write_nc_radiance`, which is the case for radiances from polar orbit satellites. We could try splitting this if statement and keep one for geo sensors and another for polar sensors. The goal was to reduce code duplication by merging these two conditions in one bigger if statement, if that makes sense.

**Comment by mgduda** [comment]
> @mgduda, that's a good point and maybe the reason for this issue. We don't need to access `geoinst_list` when `write_opt == write_nc_radiance`, which is the case for radiances from polar orbit satellites. We could try splitting this if statement and keep one for geo sensors and another for polar sensors. The goal was to reduce code duplication by merging these two conditions in one bigger if statement, if that makes sense.

I think I understand the intent -- to share code between geo and polar sensors. It looks like the `set_ahi_obserr` subroutine handles the specific instrument obs error (it is being passed `geoinst_list(ityp)` as its first argument), so perhaps this current code
```fortran
         if  ( geoinst_list(ityp) == 'ahi_himawari8' ) then
             call set_ahi_obserr(geoinst_list(ityp), xdata(ityp,itim)%nvars, obserr)
         else
             call set_brit_obserr(inst_list(ityp), xdata(ityp,itim)%nvars, obserr)
         end if
```
could become
```fortran
         if  ( write_opt == write_nc_radiance_geo ) then
             call set_ahi_obserr(geoinst_list(ityp), xdata(ityp,itim)%nvars, obserr)
         else
             call set_brit_obserr(inst_list(ityp), xdata(ityp,itim)%nvars, obserr)
         end if
```
?
